Refs #15095 - improve cache debug logging

lzap opened the pull request:
A user reports memory leak, this is an attempt to add more debugging to see if the cache is not culprit of the problem.

lzap committed:
Refs #15095 - improve cache debug logging

ekohl commented:
All tests now fail with `undefined method `[]' for nil:NilClass (NoMethodError)` and I think it's related

lzap commented:
Yeah you are right, fixed.

iNecas commented on modules/puppet_proxy_puppet_api/v3_environment_classes_api_classes_retriever.rb:
```diff
@@ -96,7 +96,10 @@ def async_get_classes(environment)
-    @m.synchronize { @futures_cache[environment] = future }
+    @m.synchronize do
+      @futures_cache[environment] = future
+      logger.debug { "Class cache: #{@classes_cache.size}, ETag cache: #{@etag_cache.size}, Futures cache: #{@futures_cache.size}" }
```
It doesn't seem the logger has an implicit prefix, so it would be good to explicitly mention where this lines are coming from: something like `"Puppet cache counts: classes %d, ETag %d,  futures %d"`

iNecas replied:
Also, I wonder, why we're only logging it at this specific place? (what about other ones where we touch any of the cache variables?)

lzap replied:
Not sure what you mean by the prefix, but rebased as you suggested.

lzap commented:
Moved the block to the future.

Also I think that the idea behind setting some values to `nil` was to remove entries from cache. But in Ruby hash implementation, setting an entry to `nil` does *not* remove the entry - it's kept with reference to `nil` value. Therefore I changed this to `delete` which does affect `count` actually. This could improve memory possibly.

iNecas reviewed (approved):
Tested and works well. Also, using delete makes sense here IMO.

iNecas commented:
@ekohl any comment before merging?

iNecas commented:
Thanks @lzap
